river/cluster/streamkmeans.py

```diff
@@ -0,0 +1,128 @@
+    in `River` uses the increamental `KMeans`. This allows the algorithm to update `KMeans`
+    without the need of re-initialization, saving a substantial amount of computing resources.
+
+    The algorithm is constructed as follows. To begin, the algorithm will be initialized
+    with an incremental `KMeans` algorithm with the same number of centers as required.
+    For a new point `p`:
+
+    * If the size of chunk is less than the maximum size allowed, add the new point to
+    the temporary chunk.
+
+    * When the size of chunk reaches the maximum value size allowed
+
+        - A new incremental `KMeans` algorithm will be initiated. This algorithm will run
+        through all points in the temporary chunk. The centers of this new algorithm will
+        be passed through the originally initialized `KMeans` to update the centers of the
+        algorithm
+
+        - All points will be deleted from the temporary chunk to continue adding new points later.
+
+    * When a prediction request arrives, the centers of the algorithm will be exactly the same
+    as the centers of the original `KMeans` at the time of retrieval.
+
+    Parameters
+    ----------
+    chunk_size
+        Maximum size allowed for the temporary data chunk.
+
+    n_clusters
+        Number of clusters generated by the algorithm.
+
+    kwargs
+        Other parameters passed to the incremental kmeans at `cluster.KMeans`.
+
+    Attributes
+    ----------
+    centers
+        Cluster centers generated from running the incremental `KMeans` algorithm
+        through centers of each chunk.
+
+    References
+    ----------
+    [^1]: O'Callaghan et al. (2002). Streaming-data algorithms for high-quality clustering.
+          In Proceedings 18th International Conference on Data Engineering, Feb 26 - March 1,
+          San Jose, CA, USA. DOI: 10.1109/ICDE.2002.994785.
+
+    Examples
+    ----------
+
+    >>> from river import cluster
+    >>> from river import stream
+
+    >>> X = [
+    ...     [1, 0.5], [1, 0.625], [1, 0.75], [1, 1.125], [1, 1.5], [1, 1.75],
+    ...     [4, 1.5], [4, 2.25], [4, 2.5], [4, 3], [4, 3.25], [4, 3.5]
+    ... ]
+
+    >>> streamkmeans = cluster.STREAMKMeans(chunk_size=3, n_clusters=2, halflife=0.5, sigma=1.5, seed=0)
+
+    >>> for x, _ in stream.iter_array(X):
+    ...     streamkmeans = streamkmeans.learn_one(x)
+
+    >>> streamkmeans.predict_one({0:1, 1:0})
+    1
+
+    >>> streamkmeans.predict_one({0:5, 1:2})
+    0
+
+    """
+
+    def __init__(self, chunk_size=10, n_clusters=2, **kwargs):
+
+        super().__init__()
+        self.time_stamp = 0
+        self.n_clusters = n_clusters
+        self.chunk_size = chunk_size
+        self.kwargs = kwargs
+
+        self._kmeans = cluster.KMeans(n_clusters=self.n_clusters, **self.kwargs)
+        self._temp_chunk = {}
+        self.centers = {}
+
+    @staticmethod
+    def _distance(point_a, point_b):
```

Comment:
This is a good chance to evaluate the need of this method. Given that it wraps the `minkowski_distsnce` utility function. We could simplify the implementation by directly using the utility function. We would need to propagate this decision into the other methods where the same method is duplicated.

Reply:
After looking again it seems that this method is not actually used. My previous comment is still valid for other methods though, but we could address that in a different PR.

Reply:
I totally agree. However, the only reason I wrap the `minkowski_distance` within `get_distance` function is to make the code cleaner and shorter. Based on the actual need of the algorithm, we can choose whether to implement this function or not. However, at this stage, I would suggest to call it in every new algorithm, whenever necessary. We can address this problem overall in one of the final commits.

Regarding this algorithm, I wil delete this `_distance` function since it is redundant.